File: C++/0116-populating-next-right-pointers-in-each-node.cpp
// Tags: Tree DFS BFS Amazon
// Time: O(n)
// Space: O(n)
// Level order traversal using queue (Bfs)
class Solution {
public:
    Node* connect(Node* root) {
        if (!root)
            return root;
        queue<Node*> q;
        q.push(root);
        while (!q.empty()) {
            int sz = q.size();
            for (int i = 0; i < sz; ++i) {
                Node* node = q.front(); q.pop();
                // point to next node on the current level, except for the last node
                node->next = i == sz - 1 ? nullptr : q.front();
                if (node->left)
                    q.push(node->left);
                if (node->right)
                    q.push(node->right);
            }
        }
        return root;
    }
};



// Time: O(n)
// Space: O(logn), due to recursion stack
// Recursive preorder tree traversal (Dfs)
class Solution {
public:
    Node* connect(Node* root) {
        // until second last level
        if (!root || !root->left)
            return root;
        // setting pointers for both children
        if (root->left)
            root->left->next = root->right;
        if (root->next)
            root->right->next = root->next->left;
        connect(root->left);
        connect(root->right);
        return root;
    }
};



// Time: O(n)
// Space: O(1)
// Iterative level order traversal
class Solution {
public:
    Node* connect(Node* root) {
        Node* head = root;
        // head is the first node of a level
        while (head) {
            Node* cur = head;
            // iterate on the current level & set next pointers of children
            // (only valid until the second last level)
            while (cur && cur->left) {
                // add link between left and right childs
                if (cur->left)
                    cur->left->next = cur->right;
                // add link between right child and left child
                // of right sibling
                if (cur->next)
                    cur->right->next = cur->next->left;
                // move to right sibling
                cur = cur->next;
            }
            // move to next level
            head = head->left;
        }
        return root;
    }
};

File: Python/0116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
# Time: O(n)
# Space: O(n)
# Iterative, level order traversal
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        que = collections.deque([root])
        while que:
            size = len(que)
            for i in range(size):
                node = que.popleft()
                node.next = que[0] if i < size - 1 else None
                if node.left and node.right:
                    que.append(node.left)
                    que.append(node.right)
        return root


# Time:  O(n)
# Space: O(1)
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        head = root
        while head:
            cur = head
            # do until the second last level
            while cur and cur.left:
                # add link between left and right childs
                cur.left.next = cur.right
                # add link between right child and left child
                # of right sibling
                if cur.next:
                    cur.right.next = cur.next.left
                cur = cur.next
            # move to next level
            head = head.left
        return root


# Time:  O(n)
# Space: O(logn), height of tree (stack)
# Recusion, dfs
class Solution:
    def connect(self, root: 'Node') -> 'Node':
            if not root or not root.left:
                return root
            if root.left:
                root.left.next = root.right
            if root.next:
                root.right.next = root.next.left
            self.connect(root.left)
            self.connect(root.right)
            return root
